modules/xfem/src/userobjects/MeshCut3DUserObject.C

```diff
@@ -0,0 +1,1106 @@
+    }
+  }
+  return elem_cut;
+}
+
+bool
+MeshCut3DUserObject::cutFragmentByGeometry(std::vector<std::vector<Point>> & /*frag_edges*/,
+                                           std::vector<CutEdge> & /*cut_edges*/,
+                                           Real /*time*/) const
+{
+  mooseError("invalid method for 3D mesh cutting");
+  return false;
+}
+
+bool
+MeshCut3DUserObject::cutFragmentByGeometry(std::vector<std::vector<Point>> & /*frag_faces*/,
+                                           std::vector<CutFace> & /*cut_faces*/,
+                                           Real /*time*/) const
+{
+  // TODO: Need this for branching in 3D
+  mooseError("cutFragmentByGeometry not yet implemented for 3D mesh cutting");
+  return false;
+}
+
+bool
+MeshCut3DUserObject::intersectWithEdge(const Point & p1,
+                                       const Point & p2,
+                                       const std::vector<Point> & vertices,
+                                       Point & pint) const
+{
+  bool has_intersection = false;
+
+  Point normal;
+  Point center;
+  findElemNormal(vertices, normal, center);
+
+  double plane_point[3] = {center(0), center(1), center(2)};
+  double planenormal[3] = {normal(0), normal(1), normal(2)};
+  double edge_point1[3] = {p1(0), p1(1), p1(2)};
+  double edge_point2[3] = {p2(0), p2(1), p2(2)};
+
+  double cut_point[3] = {0.0, 0.0, 0.0};
+
+  if (Xfem::plane_normal_line_exp_int_3d(
+          plane_point, planenormal, edge_point1, edge_point2, cut_point) == 1)
+  {
+    Point temp_p(cut_point[0], cut_point[1], cut_point[2]);
+    if (isInsideCutPlane(vertices, temp_p) && isInsideEdge(p1, p2, temp_p))
+    {
+      pint = temp_p;
+      has_intersection = true;
+    }
+  }
+  return has_intersection;
+}
+
+bool
+MeshCut3DUserObject::findIntersection(const Point & p1,
+                                      const Point & p2,
+                                      const std::vector<Point> & vertices,
+                                      Point & pint) const
+{
+  // find directional intersection
+  // along the positive extension of the vector from p1 to p2
+  bool has_intersection = false;
+
+  Point normal;
+  Point center;
+  findElemNormal(vertices, normal, center);
+
+  double plane_point[3] = {center(0), center(1), center(2)};
+  double planenormal[3] = {normal(0), normal(1), normal(2)};
+  double p_begin[3] = {p1(0), p1(1), p1(2)};
+  double p_end[3] = {p2(0), p2(1), p2(2)};
+
+  double cut_point[3] = {0.0, 0.0, 0.0};
+
+  if (Xfem::plane_normal_line_exp_int_3d(plane_point, planenormal, p_begin, p_end, cut_point) == 1)
+  {
+    Point p(cut_point[0], cut_point[1], cut_point[2]);
+    Real dotp = ((p - p1) * (p2 - p1)) / ((p2 - p1) * (p2 - p1));
+    if (isInsideCutPlane(vertices, p) && dotp > 1)
+    {
+      pint = p;
+      has_intersection = true;
+    }
+  }
+  return has_intersection;
+}
+
+bool
+MeshCut3DUserObject::isInsideEdge(const Point & p1, const Point & p2, const Point & p) const
```

Comment:
Are you sure we don't have this logic in the Geometric search methods? I'd be we do.